src/Notification/NotificationSyncer.php

```diff
@@ -215,14 +168,16 @@ protected function setDeleted(array $ids, $isDeleted)
         Notification::whereIn('id', $ids)->update(['is_deleted' => $isDeleted]);
     }
 
+
+
     /**
      * Construct an array of attributes to be stored in a notification record in
      * the database, given a notification blueprint.
      *
      * @param \Flarum\Notification\Blueprint\BlueprintInterface $blueprint
      * @return array
      */
-    protected function getAttributes(Blueprint\BlueprintInterface $blueprint)
+    public static function getAttributes(Blueprint\BlueprintInterface $blueprint)
```

Comment:
Since this method only operates on the internals of the `$blueprint` object anyway, I am thinking we could move it to the `BlueprintInterface` and properly encapsulate the data. A quick code search showed me that this is the only place where the three methods `getFromUser()`, `getSubject()` and `getData()` are used, so this tells me they are a goodcandidate to be condensed into one. Would certainly allow for more customized behavior per blueprint in the future.

The question would be how to do this in a backwards-compatible way. My suggestion:
- Add the `getAttributes()` method to each blueprint *implementation* right now. Those will probably need far less conditions, because they already know their attributes.
- Here, check for the method to exist and fall back to the current implementation (backwards compatibility for non-official extensions).
- *Only in the next release*, replace the three small methods with the new `getAttributes()` method on the interface.

That would also answer your question about re-using this static method. :wink:

Reply:
I see where you want to go, the thing is Interfaces cannot have methods with bodies, will go for an abstract then. Is that okay?

Reply:
Personally all this backward compatibility consideration is highly unnecessary, I think there is maybe one or two outside core and fof that do this. And we overcomplicate the next release with technical debt..

Reply:
My whole point was to avoid the unnecessary conditions in `getAttributes()` by inlining the methods it calls, in each class' implementation.

The three instance methods of `BlueprintInterface` will simply be collapsed into one `getAttributes()` method.

You should add this method for the `DiscussionRenamedBlueprint`, as well. If you ignore BC concerns, then you will have to do it for all bundled extensions' blueprints as well.

Reply:
Since the BC stuff worked well in previous releases, I added a small layer here, thus avoiding the abstract base class. It will be pretty simple to remove again for beta.14.